### PR TITLE
reduce usage of fatalError

### PR DESCRIPTION
**Pull request by tomerd**
motivation: less crashes

changes:
* make InternalError assert on construction
* replace fatalError with InternalError where possible
* adjust downstream code to newly throwing functions

### CODE REVIEW

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@swift-ci please smoke test

File: Sources/Basics/Errors.swift [hunk 1]
```diff
@@ -11,6 +11,7 @@
 public struct InternalError: Error {
     private let description: String
     public init(_ description: String) {
-        self.description = description
+        assertionFailure(description)
```
Comment (neonichu) [inline Sources/Basics/Errors.swift:14]: I think we need to use `preconditionFailure` if we want this to show up in release builds as well, right? https://developer.apple.com/documentation/swift/1539616-assertionfailure

**Review by neonichu — APPROVED** [review]

**Comment by neonichu** [comment]
Looks like SourceKit-LSP will need a corresponding update because it uses some of the APIs that have now become throwing.

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Review by abertelrud — APPROVED** [review]

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Review by abertelrud — APPROVED** [review]

**Pull request merged**